gcli/src/meta/executor.rs

```diff
@@ -0,0 +1,218 @@
+        linker.define("env", "gr_size", funcs::gr_size(&mut store, memory))?;
+        linker.define("env", "gr_reply", funcs::gr_reply(&mut store, memory))?;
+        linker.define("env", "gr_panic", funcs::gr_panic(&mut store, memory))?;
+        linker.define("env", "gr_oom_panic", funcs::gr_oom_panic(&mut store))?;
+        linker.define("env", "gr_out_of_gas", funcs::gr_out_of_gas(&mut store))?;
+        linker.define("env", "gr_block_height", funcs::gr_block_height(&mut store))?;
+        linker.define(
+            "env",
+            "gr_block_timestamp",
+            funcs::gr_block_timestamp(&mut store),
+        )?;
+    }
+
+    let instance = linker
+        .instantiate(&mut store, &module)
+        .unwrap()
+        .start(&mut store)?;
+
+    let metadata = instance
+        .get_export(&store, method)
+        .and_then(Extern::into_func)
+        .ok_or_else(|| anyhow!("could not find function \"{}\"", method))?
+        .typed::<(), (), _>(&mut store)?;
+
+    metadata.call(&mut store, ())?;
+    Ok(store.state().msg.clone())
+}
+
+mod funcs {
+    use super::HostState;
+    use wasmi::{
+        core::{memory_units::Pages, Trap, TrapCode},
+        AsContext, AsContextMut, Caller, Extern, Func, Memory, Store,
+    };
+
+    pub fn alloc(store: &mut Store<HostState>, memory: Memory) -> Extern {
+        Extern::Func(Func::wrap(
+            store,
+            move |mut caller: Caller<'_, HostState>, pages: i32| {
+                memory
+                    .clone()
+                    .grow(caller.as_context_mut(), Pages(pages as usize))
+                    .map_or_else(
+                        |err| {
+                            log::error!("{err:?}");
+                            u32::MAX as i32
+                        },
+                        |pages| pages.0 as i32,
+                    )
+            },
+        ))
+    }
+
+    pub fn free(ctx: impl AsContextMut) -> Extern {
+        Extern::Func(Func::wrap(ctx, |_: i32| 0))
+    }
+
+    pub fn gr_panic(ctx: &mut Store<HostState>, _memory: Memory) -> Extern {
+        Extern::Func(Func::wrap(
+            ctx,
+            move |mut _caller: Caller<'_, HostState>, _ptr: u32, _len: i32| {},
+        ))
+    }
+
+    pub fn gr_oom_panic(ctx: impl AsContextMut) -> Extern {
+        Extern::Func(Func::wrap(ctx, || {
+            log::error!("OOM panic occurred");
+            Ok(())
+        }))
+    }
+
+    pub fn gr_read(ctx: &mut Store<HostState>, memory: Memory) -> Extern {
+        Extern::Func(Func::wrap(
+            ctx,
+            move |mut caller: Caller<'_, HostState>, at: u32, len: i32, buff: i32, err: i32| {
+                let (at, len, buff, err) = (at as _, len as _, buff as _, err as _);
+
+                let msg = &caller.host_data().msg;
+                let mut payload = vec![0; len];
+                if at + len <= msg.len() {
+                    payload.copy_from_slice(&msg[at..(at + len)]);
```

Comment:
Could we alloc it inside the `if` branch? Otherwise, `payload` allocates and deallocates memory in case of an error.
```suggestion
                let payload = if at + len <= msg.len() {
                    msg[at..(at + len)].to_vec();
```